# Project 2: Looping

## Exercise 2-1: Mileage Allowance

Write a script that will allow a user to enter their monthly mileage allowance (for example, a company car or a leased car might have a mileage allowance). The user enters their weekly mileage data to determine if they're above or below their allowance.

Use a constant to hold the sentinel value of -99. Output this constant, rather than the literal "-99", when you're instructing the user what to enter.

Use a while loop. Make sure to notify the user exactly what to type in to end the loop.

Use an accumulator variable to sum the entries.

Use a counter variable to keep track of which mileage entry is being entered, number 1, 2, etc. Notice in the sample output below that the prompt is "Enter mileage entry #1" -- that '1' will be your counter variable.

At the end display if the user is over, under, or at their mileage allowance.

Sample output:

    Please enter your monthly mileage allowance in whole USD: 500
        Enter mileage entry #1 (-99 to quit): 325
        Enter mileage entry #2 (-99 to quit): 45.5
        Enter mileage entry #3 (-99 to quit): 200
        Enter mileage entry #4 (-99 to quit): -99

    The sum of mileage is: 570.5
    The allowance is: 500.0
    You are over your mileage allowance.

In [6]:
# introduce the program
input_val = 0 # initialized to 0 to start the loop at least once
input_msg = "Please enter your monthly mileage allowance (USD):"
while (input_val < 1):
    input_val = float(input(input_msg))
    print(input_msg, input_val)
allowance = round(input_val, 2) # because the smallest unit of money is a cent

# helper variables
counter = 1
quit_val = -99
total = 0
input_val = 0 # initialized to 1 to start the loop at least once
input_msg = f"\tEnter mileage entry #{counter} ({quit_val} to quit):"

# loop & add
while input_val != quit_val:
    input_val = float(input(input_msg))
    print(input_msg, input_val)
    if input_val != quit_val:
        total += input_val
total = round(total, 2) # because the smallest unit of money is a cent

# print budget info
print("\nMileage total:", total)
print("Allowance:", allowance)
if (total < allowance):
    print("You're within your budget. Congratulations!")
else:
    print("You are over your allowance")

Please enter your monthly mileage allowance (USD): 200.0
	Enter mileage entry #1 (-99 to quit): 50.0
	Enter mileage entry #1 (-99 to quit): 20.0
	Enter mileage entry #1 (-99 to quit): 76.65
	Enter mileage entry #1 (-99 to quit): -99.0

Mileage total: 146.65
Allowance: 200.0
You're within your budget. Congratulations!


## Exercise 2-2: ASCII Table

A for loop would be your use of choice. Use the proper range.

If x is an integer, you can display the ASCII-equivalent of that integer with chr(x).

Note the output formatting. You'll need to use output formatting (F-strings) to achieve the right-justified first column of data. See page 75 of the textbook, Section 2.10, 'Specifying a Minimum Field Width.'

### Sample Output

    ASCII Char
       33    !
       34    "
       35    #
       36    $
       37    %
    ...output truncated; your output will be much more
      122    z
      123    {
      124    |
      125    }
      126    ~

In [10]:
# the min & max values of printable ASCII characters
# source: https://www.ascii-code.com/
ascii_min = 33
ascii_max = 126

# prints a table
print(f'{"ASCII":>6}{"Char":>6}')
for i in range(ascii_min, ascii_max + 1):
    print(f'{i:>6}{chr(i):>6}')

 ASCII  Char
    33     !
    34     "
    35     #
    36     $
    37     %
    38     &
    39     '
    40     (
    41     )
    42     *
    43     +
    44     ,
    45     -
    46     .
    47     /
    48     0
    49     1
    50     2
    51     3
    52     4
    53     5
    54     6
    55     7
    56     8
    57     9
    58     :
    59     ;
    60     <
    61     =
    62     >
    63     ?
    64     @
    65     A
    66     B
    67     C
    68     D
    69     E
    70     F
    71     G
    72     H
    73     I
    74     J
    75     K
    76     L
    77     M
    78     N
    79     O
    80     P
    81     Q
    82     R
    83     S
    84     T
    85     U
    86     V
    87     W
    88     X
    89     Y
    90     Z
    91     [
    92     \
    93     ]
    94     ^
    95     _
    96     `
    97     a
    98     b
    99     c
   100     d
   101     e
   102     f
   103     g
   104     h
   105     i
   106     j
   107     k
   108     l

### Extra credit option [5 pts]

Note: if you are pursuing the extra credit option, please put the comment # EXTRA CREDIT OPTION at the top of your code.

Create this additional output, which outputs the ASCII table in columns of 17. Note that lots of output formatting will be required!

    ASCII:   33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49
    Char:     !   "   #   $   %   &   '   (   )   *   +   ,   -   .   /   0   1

    ASCII:   50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66
    Char:     2   3   4   5   6   7   8   9   :   ;   <   =   >   ?   @   A   B

    ASCII:   67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83
    Char:     C   D   E   F   G   H   I   J   K   L   M   N   O   P   Q   R   S

    ASCII:   84  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 100
    Char:     T   U   V   W   X   Y   Z   [   \   ]   ^   _   `   a   b   c   d

    ASCII:  101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117
    Char:     e   f   g   h   i   j   k   l   m   n   o   p   q   r   s   t   u

    ASCII:  118 119 120 121 122 123 124 125 126
    Char:     v   w   x   y   z   {   |   }   ~